### PR TITLE
Add ncclNotifyTag API for user-defined profiler-v7

### PR DESCRIPTION
**Pull request by dmonakhov**
# Add ncclNotifyTag API for user-defined profiler annotations (v7)

## Motivation

Training frameworks (PyTorch, Megatron, etc.) need a way to correlate NCCL
collective performance with application-level phases -- forward pass, backward
pass, optimizer step, data loading, etc. Today, profiler plugins see a flat
stream of collectives with no application context.

`ncclNotifyTag` lets users inject lightweight string annotations into the
profiler event stream. Plugins can then attribute bandwidth/latency metrics
to specific training phases, enabling:

- Per-phase collective performance dashboards (Prometheus/Grafana)
- Training step boundaries in Chrome traces
- Straggler detection scoped to specific communication patterns

## Original feature request
- https://github.com/NVIDIA/nccl/issues/1916

## API

```c
ncclResult_t ncclNotifyTag(const char* tag, ncclComm_t comm, cudaStream_t stream);
```

- `tag`: null-terminated string, max 32 bytes including terminator (NCCL_TAG_MAX_LEN)
- `comm`: communicator to annotate
- `stream`: unused (reserved for future use)
- Returns `ncclInvalidArgument` if `tag` or `comm` is NULL

The call is synchronous and lightweight -- no CUDA operations, no allocations.

## Usage example

```c
// Annotate training phases for profiler plugins
ncclNotifyTag("forward", comm, stream);
ncclAllReduce(sbuf, rbuf, N, ncclFloat, ncclSum, comm, stream);

ncclNotifyTag("backward", comm, stream);
ncclAllReduce(sbuf, rbuf, N, ncclFloat, ncclSum, comm, stream);
```

### Profiler plugin API v7

Extends v6 with a new event type `ncclProfileUserTag = (1 << 15)` and a new
descriptor field:

```c
struct {
  const char* tag;  // pointer valid only during startEvent/stopEvent
} userTag;
```

NCCL core delivers UserTag as a start/stop event pair. The tag pointer is valid
only for the duration of the call -- plugins must copy the string if they need
to retain it. NCCL core does zero string copying itself.

Backward compatibility: v6 and earlier plugins receive a v7 shim that filters
out UserTag events and clears the UserTag activation bit. No changes needed
in existing plugins.

### Plugin version chain

```
ncclProfiler_v7 (native)
  -> ncclProfiler_v6 (shim: filters UserTag events)
    -> ncclProfiler_v5 (shim: filters UserTag + CE events)
      -> v4 -> v3 -> v2 -> v1
```

### Example plugin (plugins/profiler/example/)

Records UserTag as Chrome trace instant events:
```json
{"name": "UserTag", "cat": "USER_TAG", "ph": "i", "s": "g",
 "args": {"tag": "forward", "rank": 0}}
```

Uses pool-based allocation (default pool size: 8, configurable via
`NCCL_PROFILE_USER_TAG_POOL_SIZE`).

### Inspector plugin (plugins/profiler/inspector/)

Captures the active tag per communicator at collective insertion time (not
completion time) for deterministic attribution. The tag appears in:

- JSON output: `"user_tag": "forward"` in `coll_perf` records
- Prometheus output: `user_tag="forward"` label on metrics

No lock needed for tag capture -- both UserTag and collective startEvent run
on the same NCCL submission thread, serialized per communicator.

## Testing 

Dockerized integration test with 13 test cases (not included in this PR):

- **Example plugin (6 tests)**: epoch tags, rapid fire (pool exhaustion),
  long tag truncation, empty tag, NULL comm, NULL tag
- **Inspector JSON (6 tests)**: same scenarios, validates `user_tag` in JSON
- **Inspector Prometheus (1 test)**: sustained collectives for 35s, validates
  `user_tag` label in .prom output

All tests pass on H200 with CUDA 13.0.

### CODE REVIEW

**Comment by gcongiu** [comment]
Thank @dmonakhov.

It seems to me like the notify API only exists to let the application talk to the profiler through NCCL. The profiler API design has the plugin form to avoid interface proliferation in NCCL. The original idea was that any new interface needed to interact with the profiler could be exposed through the profiler plugin API rather than the NCCL API.

Can't your plugin expose a tag interface directly to the application without going through NCCL? For example, a way for the application to talk to the plugin using an annotation API could be to use NVTX. I think Pytorch Kineto already uses NVTX annotations in NCCL.

**Comment by gcongiu** [comment]
@dmonakhov we are considering a different approach where the collective API calls will be extended with a configuration param (e.g., `ncclAllReduce(..., ncclConfParam_t* param)`). The param can contain the communication phase tag as attribute and correctly propagate it to the profiler without any need for stream synchronization between one phase and the next. This will require you to adopt the new APIs in your code, will this be a problem? Do you have any concerns/comments about this proposal?

**Comment by dmitry-monakhov** [comment]
@gcongiu 
Hm.. Interesting. Please correct me if I am wrong, you want   ncclConfParam_t  act as  some generic container similar to standard go-lang context  or open telemetry baggage.  Both contests proven to works.  If we can attach KV list to any nccl-op, and it will be visiable to profiler it would be awesome. So it seems like a better idea then my original proposal. Because give more flexibility for user to implement tracing at any scale. This likely result a lot of modifications in our code, but it is doable and this is for good reason.

**Comment by yfguo** [comment]
@dmitry-monakhov 
I would like to follow up on the proposed interface design. We are thinking something like this. There can be a `ncclCollConfig_t` struct which contains things that can be set at per-collective call level. If we add the uint64_t field for user tag, we can "tag" a specific collective call.
```cpp
ncclCollConfig_t forward_pass_cfg = NCCL_COLL_CONFIG_INITIALIZER;
ncclCollConfig_t backward_pass_cfg = NCCL_COLL_CONFIG_INITIALIZER;
forward_pass_cfg.userProfilerTag = FORWARD_TAG;
backward_pass_cfg.userProfilerTag = BACKWARD_TAG;

// run allreduce for forward pass
ncclAllReduceConfig(sbuf, rbuf, N, ncclFloat, ncclSum, comm, stream, &forward_pass_cfg);
// same cfg can be reused for other calls
ncclAllGatherConfig(..., &forward_pass_cfg);
// regular collective call (without config) is not tagged.
ncclAllReduce(...); // run as untagged

// run allreduce for backward pass
ncclAllReduceConfig(sbuf, rbuf, N, ncclFloat, ncclSum, comm, stream, &backward_pass_cfg);
```
Do you think this would satisfy your use case?

**Comment by dmonakhov** [comment]
> @dmitry-monakhov I would like to follow up on the proposed interface design. We are thinking something like this. There can be a `ncclCollConfig_t` struct which contains things that can be set at per-collective call level. If we add the uint64_t field for user tag, we can "tag" a specific collective call.
> 
> ```c++
> ncclCollConfig_t forward_pass_cfg = NCCL_COLL_CONFIG_INITIALIZER;
> ncclCollConfig_t backward_pass_cfg = NCCL_COLL_CONFIG_INITIALIZER;
> forward_pass_cfg.userProfilerTag = FORWARD_TAG;
> backward_pass_cfg.userProfilerTag = BACKWARD_TAG;
> 
> // run allreduce for forward pass
> ncclAllReduceConfig(sbuf, rbuf, N, ncclFloat, ncclSum, comm, stream, &forward_pass_cfg);
> // same cfg can be reused for other calls
> ncclAllGatherConfig(..., &forward_pass_cfg);
> // regular collective call (without config) is not tagged.
> ncclAllReduce(...); // run as untagged
> 
> // run allreduce for backward pass
> ncclAllReduceConfig(sbuf, rbuf, N, ncclFloat, ncclSum, comm, stream, &backward_pass_cfg);
> ```
> 
> Do you think this would satisfy your use case?

@yfguo

Ack. It will works,   uint64_t is wide enough to implement any encoding scheme.
Your approach indeed cleaner, because of explicit parameter is passed. 
From integration point it require some ground work, as majority of users use NCCL indirectly via pytorch nowadays :). It seems it is reasonably straight forward.  We can use existing thread-local trick like https://github.com/pytorch/pytorch/blob/main/c10/util/ThreadLocalDebugInfo.h#L61 to set UserProfilerTagDebugInfo , and all later ill be used during collective calls

**Comment by yfguo** [comment]
@dmitry-monakhov The user profiler tag changes have landed in the `dev` branch as these commits:
69e7c2c0c06f7744d7bde0c24e566a2b2ec84b99
0dab26429c5a0ef15339fe0285809721b67302bf
deb1f27a074afb3eb65106419168814655624563

Please give it a try.

**Comment by xiaofanl-nvidia** [comment]
This will be released as part of 2.31 shortly. Closing.

@dmonakhov please let us know if you see any issues when using this feature! Thank you for your RFE, which motivated us to add the collConfig feature.
